traits/trait_types.py

```diff
@@ -3417,7 +3420,7 @@ def validate(self, object, name, value):
             self.validate_failed(object, name, value)
         else:
             result = self.default_value
-            if isinstance(result, _InstanceArgs):
```

Comment:
It turns out that this pre-PR version of the code was actually buggy: in the callable case, `self.default_value` could already have been replaced if `get_default_value` had been invoked. The tests added in this PR fail with this version of the code.